src/Sema.zig

```diff
@@ -39127,6 +38612,1032 @@ pub fn resolveDeclaredEnum(
+        sema.pt.zcu.intern_pool.indexToKey(Type.toIntern(src_ptr_ty)).ptr_type;
+
+    // *T   ;; Used to refer to the element at the sentinel index.
+    const elem_ptr_ty: Type = blk: {
+        var elem_ptr_ty_key: InternPool.Key.PtrType = ptr_ty_key;
+        elem_ptr_ty_key.child = Type.toIntern(elem_ty);
+        elem_ptr_ty_key.flags.size = .One;
+        break :blk try sema.pt.ptrTypeSema(elem_ptr_ty_key);
+    };
+
+    // [*]T ;; Used to compute the start pointer by pointer arithmetic.
+    const many_ptr_ty: Type = blk: {
+        var many_ptr_ty_key: InternPool.Key.PtrType = ptr_ty_key;
+        many_ptr_ty_key.child = Type.toIntern(elem_ty);
+        many_ptr_ty_key.flags.size = .Many;
+        break :blk try sema.pt.ptrTypeSema(many_ptr_ty_key);
+    };
+
+    //
+    // COMPUTE AND RESOLVE VALUES:
+    //
+
+    // Initial state of values and safety checks (4 bytes):
+    var sa: SliceAnalysis = .{};
+
+    // Attempt to resolve the input pointer:
+    const src_ptr_val: Value = val: {
+        if (try sema.resolveValue(src_ptr)) |val| {
+            // The pointer operand is not optional--`unknown` means the input
+            // pointer is undefined.
+            sa.src_ptr = if (val.isUndef(sema.pt.zcu)) .unknown else .known;
+            break :val val;
+        }
+        break :val Value.undef;
+    };
+
+    // Try hard to find the actual size of any `comptime` memory. It is easy to
+    // fabricate a fake size by casting any `comptime`-known pointer to an array
+    // which can then coerce to a slice.
+    const src_mem_len: Air.Inst.Ref = inst: {
+        if (feature.prevent_inval_ptr and sa.src_ptr == .known) {
+            const elem_size: u64 = try elem_ty.abiSizeSema(sema.pt);
+            const ptr_val: Value = if (src_ptr_ty_size == .Slice)
+                Value.slicePtr(src_ptr_val, sema.pt.zcu)
+            else
+                src_ptr_val;
+            // Do not rely on weird kinds of pointers if there is no explicit
+            // length to compare it against. Reject reinterpreted `eu_payload`
+            // and `opt_payload` pointers.
+            const want_error: bool = !src_ptr_explicit_len;
+            if (elem_size == 0) {
+                // Only best-effort to give users what they ask for when it is
+                // not an unambiguous disaster.
+                if (try sema.elemLenOfContainingDecl(block, src, ptr_val, elem_ty, want_error)) |idx_int| {
+                    const len: usize = try sema.usizeCast(block, src, idx_int);
+                    break :inst try sema.pt.intRef(Type.usize, len);
+                }
+            } else {
+                if (try sema.bytesOfContainingDecl(block, src, ptr_val, want_error)) |memsz_int| {
+                    const len: usize = try sema.usizeCast(block, src, memsz_int / elem_size);
+                    break :inst try sema.pt.intRef(Type.usize, len);
+                }
+            }
+        }
+        break :inst .none;
+    };
```

Comment:
Take a look at the recent closure of #19797 -- apologies for this, myself and Andrew both previously failed to look closely at that issue. Being able to construct invalid pointers and slices is permitted in Zig (unless they are sliced with a sentinel, in which case it is Unchecked Illegal Behaviour or a compile error if the sentinel address isn't valid). Please remove this check.

Reply:
> Being able to construct invalid pointers and slices is permitted in Zig
Why *should it* be permitted for comptime-known memory?

Reply:
Because it is permitted for runtime memory. The reason you can run arbitrary functions at compile-time is precisely because it is the same language in both cases.

Reply:
It is an option for a reason. It can be disabled. But why would that ever be desirable within the language?

There is no reason to delay diagnosing this error.

Reply:
Because it's not an error in Zig's memory model to construct invalid pointers, only to load or store through them.

Reply:
This only applies to slices afaik, not pointers

Reply:
For all intents and purposes you should think of `[:x]attrs T` as being equivalent to `struct { ptr: [*:x]attrs T, len: usize }`.

Reply:
Constructing a slice pointing past the last element is possible, but you have to do some extra effort to avoid bounds checks. And if invalid slices are fine, why even have bounds checks?

Reply:
Because they can be trivially safety-checked at runtime, and it is useful to do so.

Reply:
> For example, do you think it should be illegal behavior to construct a pointer to the end of an array, pointing 1 past the last element? If not, why not?

It should be illegal if that pointer is outside the compile-time allocation for the containing declaration, because that pointer can not correspond to well-defined memory if it escapes to runtime. That pointer means nothing when the compiler allows it to be created, and only gains any value by the chance arrangement of the file.

This check only prevents the user from creating pointers which would definitely produce compile errors if they were accessed at compile time, and would definitely produce undefined behaviour if they were accessed at runtime. There is no potential usage for such pointers which are always undefined.

Reply:
Is the concern that something like this will not work?

```zig
comptime {
    var buf: [2][]const u8 = .{ "a", "b" };
    const end: [*][]const u8 = buf[2..];
    var start: [*][]const u8 = buf[0..];
    while (start != end) {
        @compileLog(start[0]);
        start += 1;
    }
}
```

If the slice has a length of zero, then it is not undefined behaviour for the pointer part to be out-of-bounds. There is no special exception for this in the implementation or in the checks being discussed here.

```
one_past_the_end.zig:6:9: error: found compile log statement
        @compileLog(start[0]);
        ^~~~~~~~~~~~~~~~~~~~~

Compile Log Output:
@as([]const u8, "a"[0..1])
@as([]const u8, "b"[0..1])
```

Reply:
Note, the syntax has produced a sound pointer because the return value has a length equal to zero, meaning no access is well-defined. What the user does with that pointer is their choice. It is not the responsibility of the syntax to prevent subsequent usage.

This check is only aimed at making slices consistently safe. If what you are describing was not permitted the compiler would not be able to build, let alone pass all tests.

Reply:
> Is the concern that something like this will not work?

No, the concern is obeying the language semantics.

Compile errors should match illegal behavior. This is important for comptime execution to work. The question, then, is whether slicing past valid memory is Illegal Behavior.

This isn't a firm rule, but generally speaking, there are two main paths to introducing illegal behavior:
* Illegal behavior *may* be introduced if it can be trivially safety-checked.
* Illegal behavior *should* be introduced if one or more targets require it.

Accessing an invalid pointer is necessarily illegal behavior by the second point: invalid pointers may not refer to mapped memory, so accessing them could cause a segfault. So, accessing an invalid pointer has to be Unchecked Illegal Behavior.

On the other hand, *constructing* an invalid pointer (and the same here applies to slices) can't be considered Illegal Behavior under either of those justifications. It's perfectly valid to do on all targets, so the second point doesn't justify it; and we can't safety-check it without an unreasonable amount of work (e.g. loading from every single created pointer). The latter point is actually much worse for slices, as such a safety check would require loading every element of the slice (it's not sufficient to just load the first and last, since you may have crossed over an invalid-memory boundary). Note that it would also violate language semantics to have such a safety check on a `volatile` pointer.

---

Regardless, this PR is not the place to bikeshed this part of the language design. I consider this matter fairly settled; please make the requested change.

Reply:
But it's already not valid to _construct_ an invalid slice through slicing syntax in many cases, because it has compile time and runtime checks on it. This check only increases the safety of slicing syntax, which already attempts to prevent invalid slices from being constructed. Constructing invalid slices through other means is still valid.

Why shouldn't `([_]u8{})[0..1000]` also compile with no errors? After all constructing invalid slices / pointers is not illegal behavior.

Reply:
> Because [slicing] can be trivially safety-checked at runtime, and it is useful to do so.

Reply:
I think it may have sunken in for me...

So if I understand the philosophy correctly, a function like:
```zig
fn alignToNthNextItem(mem: [*]u8, n: usize, T: type) *T
```

Which will align forwards to match T's alignment, then move forwards by n times T's in mem to produce a pointer address to the nth aligned T.

And usage code like:
```zig
var mem = [1]u8{0};
const runtime_ptr = alignToNthNextItem(&mem, 1000, u128);
const comptime_ptr = comptime alignToNthNextItem(&mem, 1000, u128);
```

Then calling the function at comptime rather than runtime should cause no additional errors, because only then can you uphold the guarantees of comptime code executing as runtime code (in the abscense of disallowed specific things like inline asm).

So when a runtime check cannot be practically implemented, no compile time check must be implemented because it would prevent comptime execution of an otherwise valid function (which may also even come from translated C).

Am I getting it correctly?